### DISCUSSION TITLE
A simpler way to add logic for filtering with logical operators

### DISCUSSION THREAD

**Discussion by frederikhors** [async-graphql/async-graphql#938] [Q&A]
For now I'm using GraphQL in Golang with [gqlgen](https://github.com/99designs/gqlgen) and [Bun](https://github.com/uptrace/bun) as ORM.

I would like to use Rust in the near future and imagine using [Async-Graphql](https://github.com/async-graphql/async-graphql) and [SeaORM](https://github.com/SeaQL/sea-orm).

I'm using [DDD](https://en.wikipedia.org/wiki/Domain-driven_design) and [Clean architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html), meaning that I'm avoiding `one-size-fits-all/lock-in` solutions like Hasura, Postgraphile, DGraph and everything like that; I can (and I eventually will, no hurry) write and organize my code in such a way to achieve their features.

For now I'm wondering about complex queries with filters like: `AND`, `OR`, `NOT` ([better described here](https://github.com/graphile-contrib/postgraphile-plugin-connection-filter/blob/master/docs/examples.md)).

Right now I need to think/write/test code all by myself and this certainly does not speed up production of the software I want to write.

Also, my code will almost certainly be slow, buggy and possibly all wrong as soon as I realize that I will need something extra for this feature.

_After all, this is true for all software and is one of the reasons why open source was born._

After some research I noticed there are two approaches to solving the problem of GraphQL data filtering:

- create all possible combinations for each field of the model ([this is how ent behaves](https://entgo.io/blog/2021/07/01/automatic-graphql-filter-generation) and [other users suggest](https://github.com/99designs/gqlgen/issues/385#issuecomment-432130569))

- create input types that contain the model fields but also logical comparators as a possible value (like [Hasura, Postgraphile and others do](https://hasura.io/docs/latest/graphql/core/databases/postgres/queries/query-filters/))

### QUESTIONS

1. What do you think about? Pros and cons of each choice?

1. If I wanna use `async-graphql` is it possible to have an easier way to map our Graphql structs to our domain code without manually write that logic by ourself?

1. Has anyone faced this problem before?

**Comment by cosmincatalin** [comment]
In HotChocolate there is built-in filtering: https://chillicream.com/docs/hotchocolate/v13/fetching-data/filtering.
A similar feature would be really nice in async-graphql.
